File: gops/sdrrcv.go
// sdrrcv.go : SDR receiver functions (Go version)
// Translated from sdrrcv.c
package main
// Buffer size constants (match original C logic)

import (
	"os"
	"sync"
)

var buffmtx sync.Mutex
var readmtx sync.Mutex
var err error // global error variable for fallback returns

// SDR receiver initialization
func RcvInit(ini *SdrIni, sdrstat *SdrStat) error {
	sdrstat.Buff = nil
	sdrstat.TmpBuff = nil
	switch ini.Fend {
	case FEND_FILE:
		var err error
		ini.fp, err = os.Open(ini.File)
		if err != nil {
			SDRPRINTF("error: failed to open file: %s\n", ini.File)
			return err
		}
		sdrstat.FendBuffSize = FILE_BUFFSIZE
		sdrstat.BuffSize = FILE_BUFFSIZE * MEMBUFFLEN
		if ini.fp != nil {
			// Allocate buffer for all channels and all data
			sdrstat.Buff = make([]byte, ini.Dtype[0]*sdrstat.BuffSize)
			if sdrstat.Buff == nil || len(sdrstat.Buff) == 0 {
				SDRPRINTF("error: failed to allocate memory for the buffer\n")
				return err
			}
		}
	default:
		return err
	}
	return nil
}

// SDR receiver quit
func RcvQuit(ini *SdrIni, sdrstat *SdrStat) error {
	switch ini.Fend {
	case FEND_FILE:
		if ini.fp != nil {
			ini.fp.Close()
			ini.fp = nil
		}
	default:
		return err
	}
	if sdrstat.Buff != nil {
		sdrstat.Buff = nil
	}
	if sdrstat.TmpBuff != nil {
		sdrstat.TmpBuff = nil
	}
	return nil
}

// Start grabber of front end (tu: tylko plik, zwraca 0)
func RcvGrabStart(ini *SdrIni) int {
	switch ini.Fend {
	default:
		return 0
	}
	return 0
}

// Push data to memory buffer from front end (tu: tylko plik)
func RcvGrabData(ini *SdrIni, sdrstat *SdrStat) int {
	switch ini.Fend {
	case FEND_FILE:
		FilePushToMemBuf(ini, sdrstat)
		Sleepms(5)
	default:
		return -1
	}
	return 0
}

// Get current data buffer from memory buffer
func RcvGetBuff(ini *SdrIni, buffloc uint64, n, ftype, dtype int, sdrstat *SdrStat, expbuf []byte) int {
	switch ini.Fend {
	case FEND_FILE:
		FileGetBuff(buffloc, n, ftype, dtype, expbuf)
	default:
		return -1
	}
	return 0
}

func FilePushToMemBuf(ini *SdrIni, sdrstat *SdrStat) {
	var nread int
	buffmtx.Lock()
	if ini.fp != nil {
		buf := make([]byte, ini.Dtype[0]*FILE_BUFFSIZE)
		n, err := ini.fp.Read(buf)
		nread = n
		if err != nil && err.Error() != "EOF" {
			SDRPRINTF("FilePushToMemBuf: read error: %v\n", err)
		}
		copy(sdrstat.Buff[int((sdrstat.BuffCnt%uint64(MEMBUFFLEN))*uint64(ini.Dtype[0])*uint64(FILE_BUFFSIZE)):], buf[:nread])
		SDRPRINTF("FilePushToMemBuf: BuffCnt=%d, bytes read=%d\n", sdrstat.BuffCnt, nread)
	}
	buffmtx.Unlock()
	if ini.fp != nil && nread < ini.Dtype[0]*FILE_BUFFSIZE {
		sdrstat.StopFlag = ON
		SDRPRINTF("end of file!\n")
	}
	readmtx.Lock()
	sdrstat.BuffCnt++
	readmtx.Unlock()
}

func FileGetBuff(buffloc uint64, n, ftype, dtype int, expbuf []byte) {
	membuffloc := dtype*int(buffloc) % (MEMBUFFLEN*dtype*FILE_BUFFSIZE)
	n = dtype * n
	nout := (membuffloc + n) - (MEMBUFFLEN * dtype * FILE_BUFFSIZE)
	bufflen := MEMBUFFLEN * dtype * FILE_BUFFSIZE
	buffmtx.Lock()
	if nout > 0 {
		start := membuffloc
		end := membuffloc + n - nout
		if start < 0 { start = 0 }
		if end < 0 { end = 0 }
		if end > bufflen { end = bufflen }
		copy(expbuf, sdrstat.Buff[start:end])
		copy(expbuf[n-nout:], sdrstat.Buff[:nout])
	} else {
		start := membuffloc
		end := membuffloc + n
		if start < 0 { start = 0 }
		if end < 0 { end = 0 }
		if end > bufflen { end = bufflen }
		copy(expbuf, sdrstat.Buff[start:end])
	}
	buffmtx.Unlock()
}

File: GpsJammerApp/backend/sdrrcv.c
#include "sdr.h"

extern int rcvinit(sdrini_t *ini) {
    fftwf_init_threads();

    sdrstat.buff = sdrstat.tmpbuff = NULL;

    if ((ini->fp = fopen(ini->file, "rb")) == NULL) {
        SDRPRINTF("error: failed to open file: %s\n", ini->file);
        return -1;
    }

    sdrstat.fendbuffsize = FILE_BUFFSIZE;
    sdrstat.buffsize = 2 * FILE_BUFFSIZE * MEMBUFFLEN;

    if (ini->fp != NULL) {
        sdrstat.buff = (uint8_t *)malloc(ini->dtype[0] * sdrstat.buffsize);
        if (NULL == sdrstat.buff) {
            SDRPRINTF("error: failed to allocate memory for the buffer\n");
            return -1;
        }
    }

    return 0;
}

extern int rcvquit(sdrini_t *ini) {

    if (ini->fp != NULL) {
        fclose(ini->fp);
        ini->fp = NULL;
    }

    if (NULL != sdrstat.buff) {
        free(sdrstat.buff);
        sdrstat.buff = NULL;
    }
    if (NULL != sdrstat.tmpbuff) {
        free(sdrstat.tmpbuff);
        sdrstat.tmpbuff = NULL;
    }
    return 0;
}

extern int rcvgrabstart(sdrini_t *ini) { return 0; }

extern int rcvgrabdata(sdrini_t *ini) {

    file_pushtomembuf();
    sleepms(5);

    return 0;
}

extern int rcvgetbuff(sdrini_t *ini, uint64_t buffloc, int n, int ftype,
                      int dtype, char *expbuf) {
    file_getbuff(buffloc, n, ftype, dtype, expbuf);
    return 0;
}

extern void file_pushtomembuf(void) {
    size_t nread = 0;

    mlock(hbuffmtx);
    if (sdrini.fp != NULL) {
        nread = fread(&sdrstat.buff[(sdrstat.buffcnt % MEMBUFFLEN) *
                                    sdrini.dtype[0] * FILE_BUFFSIZE],
                      1, sdrini.dtype[0] * FILE_BUFFSIZE, sdrini.fp);
    }
    unmlock(hbuffmtx);

    if ((sdrini.fp != NULL && (int)nread < sdrini.dtype[0] * FILE_BUFFSIZE)) {
        sdrstat.stopflag = ON;
        SDRPRINTF("end of file!\n");
    }

    mlock(hreadmtx);
    sdrstat.buffcnt++;
    unmlock(hreadmtx);
}

extern void file_getbuff(uint64_t buffloc, int n, int ftype, int dtype,
                         char *expbuf) {
    uint64_t membuffloc =
        dtype * buffloc % (MEMBUFFLEN * dtype * FILE_BUFFSIZE);
    int nout;
    int i;
    int total_bytes;

    n = dtype * n;
    total_bytes = n;
    nout = (int)((membuffloc + n) - (MEMBUFFLEN * dtype * FILE_BUFFSIZE));

    mlock(hbuffmtx);

    if (nout > 0) {
        memcpy(expbuf, &sdrstat.buff[membuffloc], n - nout);
        memcpy(&expbuf[(n - nout)], &sdrstat.buff[0], nout);
    } else {
        memcpy(expbuf, &sdrstat.buff[membuffloc], n);
    }
    unmlock(hbuffmtx);

    for (i = 0; i < total_bytes; i++) {
        expbuf[i] = (char)((unsigned char)expbuf[i] - 128);
    }
}

